src/helpers/map.js

```diff
@@ -169,3 +173,163 @@ export function findTile(tiles) {
   console.error(`Unable to find tile: ${tiles}`);
   return {};
 }
+
+
+/*
+ * Initializes and returns a 2D array with the specified width, height, and
+ * default value.
+ *
+ * width: the width of the initialized 2D array
+ * height: the height of the initialized 2D array
+ * defaultVal: the value to give each element in the initialized 2D array
+ */
+export function init2dArray(width, height, defaultVal) {
+  const matrix = [];
+
+  for (let x = 0; x < width; x++) {
+    matrix[x] = new Array(height);
+    for (let y = 0; y < height; y++) {
+      matrix[x][y] = defaultVal;
+    }
+  }
+
+  return matrix;
+}
+
+
+/*
+ * Returns the sum of all corresponding elements from two matrices being
+ * multiplied together. For example:
+ *   a = 1 2   b = 4 1
+ *       3 4       2 3
+ *   multiplyCorrespondingElementsAndSum(a, b)
+ *   (1*4) + (2*1) + (3*2) + (4*3) = 21
+ *
+ * m1: the first matrix
+ * m2: the second matrix (of the same dimensions as m1)
+ */
+export function multiplyCorrespondingElementsAndSum(m1, m2) {
+  let condition = (m1.length === m2.length);
+  let errorMessage = 'multiplyCorrespondingElementsAndSum: m1 and m2 are not the same width';
+  utils.assert(condition, errorMessage);
+
+  condition = (m1[0].length === m2[0].length);
+  errorMessage = 'multiplyCorrespondingElementsAndSum: m1 and m2 are not the same height';
+  utils.assert(condition, errorMessage);
+
+  const mWidth = m1.length;
+  const mHeight = m1[0].length;
+  let sum = 0;
+
+  for (let x = 0; x < mWidth; x++) {
+    for (let y = 0; y < mHeight; y++) {
+      sum += m1[x][y] * m2[x][y];
+    }
+  }
+
+  return sum;
+}
+
+
+/* Returns a matrix with a buffer applied around its perimeter. Example:
+ *   Take the element values of the matrix and apply a bufSize perimeter
+ *   of value bufVal.
+ *     mWithBuf = 1 2 3   bufSize = 2   mWithBuf = 1 1 1 1 1 1 1
+ *                4 5 6                            1 1 1 1 1 1 1
+ *                7 8 9   bufVal = 1               1 1 1 2 3 1 1
+ *                                                 1 1 4 5 6 1 1
+ *                                                 1 1 7 8 9 1 1
+ *                                                 1 1 1 1 1 1 1
+ *                                                 1 1 1 1 1 1 1
+ *
+ * m: a 2D array, which will have a buffer added around its perimeter
+ * bufSize: the size of buffer to add around the matrix
+ * bufVal: the value to fill the buffer with
+ */
+export function addBufferTo2dArray(m, bufSize, bufVal) {
+  const mWidth = m.length;
+  const mHeight = m[0].length;
+  const mWithBufWidth = mWidth + (bufSize * 2);
+  const mWithBufHeight = mHeight + (bufSize * 2);
+
+  // Create a grid with matrix's values and a bufSize perimeter of zeros
+  const mWithBuf = init2dArray(mWithBufWidth, mWithBufHeight, bufVal);
+  fillGridWithSubgrid(mWithBuf, m, bufSize, bufSize);
```

Comment:
does this work with consts? We shouldn't be able to modify `mWithBuf` if it's a const